hbase-server/src/main/java/org/apache/hadoop/hbase/ipc/ServerCall.java

```diff
@@ -97,11 +97,13 @@
   private long exceptionSize = 0;
   private final boolean retryImmediatelySupported;
 
-  // This is a dirty hack to address HBASE-22539. The lowest bit is for normal rpc cleanup, and the
-  // second bit is for WAL reference. We can only call release if both of them are zero. The reason
-  // why we can not use a general reference counting is that, we may call cleanup multiple times in
-  // the current implementation. We should fix this in the future.
-  private final AtomicInteger reference = new AtomicInteger(0b01);
+  // This is a dirty hack to address HBASE-22539. The highest bit is for rpc ref and cleanup, and
+  // the rest of the bits are for WAL reference count. We can only call release if all of them are
+  // zero. The reason why we can not use a general reference counting is that, we may call cleanup
+  // multiple times in the current implementation. We should fix this in the future.
+  // The refCount here will start as 0x80000000 and increment with every WAL reference and decrement
+  // from WAL side on release
+  private final AtomicInteger reference = new AtomicInteger(0x80000000);
```

Comment:
We can represent this as an atomic boolean and atomic integer together, right? We are either checking or adjusting the high order bit atomically or adjusting the refcount atomically, we aren't doing both together. There isn't a case where the high order bit is updated *and* the refcount is changed and both must be applied as a single atomic operation. 

The code will be much more understandable and maintainable with a well-named atomic boolean field and the result is no longer a dirty hack.

Reply:
Like this idea. This new class needs to be generic as not only in this rpc handler/wal subsystem needs to coordinate. The replication path also needs similar coordination.

Reply:
Right - as @Apache9 mentions above, the check whether rpc ref is false AND wal reference count = 0 needs to happen atomically and make sure the ByteBuffer is released exactly once. Thus single atomicinteger mechanism provides synchronized way of dealing with this @anoopsjohn FYI

Reply:
I see, it's fine, if I misread the code. 

I am guilty myself of employing this trick, where a high order bit of a counter is overloaded to be a flag. I feel it leads to less maintainable code than the alternative, but if it has to be done it has to be done. The issue is someone less familiar with the code than the current author and reviewers might forget to mask when reading or updating. 

> Maybe we could introduce a separated class for this logic to make the code in ServerCall cleaner, and add clear document for the new class about the logic.

This is a good suggestion.